Add more exception types (not just raising `BaseException`).

JoeZiminski opened the pull request:
closes #246. It is useful to have a central function in which all errors are raised, for example if needing to tear down any functionality (e.g. close loggers) before raising an error.

Previously this centralisation meant that only `BaseException` was raised. Now, the central error-raising functions take the type of exception to raise, and raise this. These means API users can catch specific exception types.

This PR also introduces two custom exceptions, `NeuroBlueprintError` (related to violations of NeuroBlueprint standard) and `ConfigError` (related to user passing bad configs).

Updated relevant tests, docs not required.

IgorTatarnikov commented on datashuttle/utils/custom_exceptions.py:
```diff
@@ -0,0 +1,6 @@
+class ConfigError(BaseException):
```
```suggestion
class ConfigError(Exception):
```
Someone might try to catch the exception using:
```
try:
    something
except Exception as e:
    cleanup
```
And not catch the exception if it inherits from `BaseException`.

IgorTatarnikov commented on datashuttle/utils/custom_exceptions.py:
```diff
@@ -0,0 +1,6 @@
+    pass
+
+
+class NeuroBlueprintError(BaseException):
```
```suggestion
class NeuroBlueprintError(Exception):
```
Someone might try to catch the exception using:
```
try:
    something
except Exception as e:
    cleanup
```
And not catch the exception if it inherits from `BaseException`.

IgorTatarnikov reviewed (commented):
Looks great! My only comment relates to the custom exceptions inheriting from `BaseException`. Could that be changed to `Exception` instead without downstream consequences? 

Also I noticed that sometime you use relative imports instead of absolute, is there a benefit or is that just a style choice? e.g.
[https://github.com/neuroinformatics-unit/datashuttle/blob/570d4d6566e343b1d1a9ed0d85df24fd32a9acd6/datashuttle/utils/folders.py#L15](url)

JoeZiminski commented:
Thanks @IgorTatarnikov! That's much better to inherit from `Exception` cheers. 

Also good point about imports, I always thought they were preferred based on nothing, [I see](https://stackoverflow.com/questions/4209641/absolute-vs-explicit-relative-import-of-python-module) absolute is much preffered! which makes sense. On [a commit](https://stackoverflow.com/questions/4209641/absolute-vs-explicit-relative-import-of-python-module) in another refactoring PR I've changed all imports to absolute and added a `ruff` rule to error on relative imports. Thanks!

IgorTatarnikov reviewed (approved):
Looks good!
